ConfigurationSubstitutor/ConfigurationSubstitutor.cs

```diff
@@ -94,12 +96,18 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
 
                         return ApplySubstitution(configuration, value, recursionDetectionSet);
                     }
-
                 }
+
                 if (substitutedValue == null && _exceptionOnMissingVariables)
                 {
                     throw new UndefinedConfigVariableException($"{_startsWith}{capture.Value}{_endsWith}");
                 }
+
+                if (substitutedValue == null && _keepNonResolvedVariables)
```

Comment:
Might be better to introduce a breaking change, we start to have to have too many flags.
Instead of having both exceptionOnMissingVariables and keepNonResolvedVariables, I would rather something like:
```c#
public enum UnresolvedVariableBehaviour {
    // UndefinedConfigVariableException is thrown for unresolved variables
    Throw,
    // Unresolved variables are ignored and replaced by empty content
    IgnorePattern,
    // The pattern for substitution is kept in case of unresolved variables 
    KeepPattern
}
```

Then the default behaviour would remain to throw and there's the possibility to Ignore (currently that's exceptionOnMissingVariables:false, or keep the variable pattern)

Reply:
I added new signatures and tagged old method as obsolete.
I kept all previous unit test to guarantee that deprecated methods are still OK.

Reply:
Thanks @MickaJsd we are almost there, good job!